etcdserver/apply.go

```diff
@@ -175,6 +175,10 @@ func (a *applierV3backend) Put(txnID int64, p *pb.PutRequest) (*pb.PutResponse,
 		}
 	}
 
+	if p.PreserveValue && rr != nil && len(rr.KVs) != 0 {
+		p.Value = rr.KVs[0].Value
+	}
```

Comment:
`p.PreserveValue && (rr == nil || len(rr.Kvs) == 0)` should be an error since it shouldn't create a new key. Txns will need a special case to check that all keys exist before doing the puts.